vowpalwabbit/config/options_cli.cc

```diff
@@ -0,0 +1,450 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "config/options_cli.h"
+
+#include <algorithm>
+#include <cctype>
+#include <iterator>
+#include <memory>
+#include <set>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <queue>
+
+#include "config/option.h"
+#include "config/options.h"
+
+#include "vw_exception.h"
+#include "vw_string_view.h"
+
+using namespace VW::config;
+
+bool is_number(const std::string& str)
+{
+  if (str.empty()) { return false; }
+
+  char* ptr;
+  std::strtof(str.c_str(), &ptr);
+  return (*ptr) == '\0';
+}
+
+bool is_number(VW::string_view str) { return is_number(std::string{str}); }

```

Comment:
Can we put the implementation here and remove the `const std::string& str` overload since it's only used as an intermediary

Reply:
This is actually necessary. Tokens are stored as string_views which are *not* null terminated, `strtof` requires a null terminated string. So the overload is actually creating a temporary just to facilitate this call. It's a bit gross I agree